impactcommon/exogenous_economy/gdppc.py

```diff
@@ -1,55 +1,213 @@
-        self.df_baseline_anyiam = df.loc[(df.scenario == ssp) & (df.year == baseline_year)].groupby('iso').median()
-        self.baseline_global = df.loc[(df.scenario == ssp) & (df.year == baseline_year)].median()
+    out = BestGDPpcProvider(
+        iam=iam,
+        ssp=ssp,
+        df_baseline=df,
+        df_growth=df_growth,
+        df_nightlights=df_nightlights,
+        startyear=startyear,
+        stopyear=stopyear
+    )
+    return out
 
-        # Load the growth rates, and split by priority of data
-        df_growth = metacsv.read_csv(files.sharedpath(growth_filepath))
-        df_growth['yearindex'] = np.int_((df_growth.year - baseline_year) / 5)
-        self.df_growth_this = df_growth.loc[(df_growth.model == iam) & (df_growth.scenario == ssp)]
-        self.df_growth_anyiam = df_growth.loc[(df_growth.scenario == ssp)].groupby(['iso', 'year']).median()
-        self.growth_global = df_growth.loc[(df_growth.scenario == ssp) & (df_growth.model == iam)].groupby(['year']).median()
 
-        # Load the nightlights
-        self.df_nightlights = metacsv.read_csv(files.sharedpath(nightlights_filepath))
+def GDPpcProvider(iam, ssp, baseline_year=2010, growth_filepath='social/baselines/gdppc-growth.csv',
+                  baseline_filepath='social/baselines/gdppc-merged-nohier.csv',
+                  nightlights_filepath='social/baselines/nightlight_weight_normalized.csv', stopyear=2100):
+    """Get BestGDPpcProvider through the legacy GDPpcProvider interface
+
+    This interface is deprecated, please use read_bestgdppcprovider() or
+    instantiate BestGDPpcProvider, directly.
+
+    Parameters
+    ----------
+    iam : str
+    ssp : str
+    baseline_year : int, optional
+    growth_filepath : str, optional
+    baseline_filepath : str, optional
+    nightlights_filepath : str, optional
+    stopyear : int, optional
+
+    Returns
+    -------
+    out : impactcommmon.exogenous_economy.BestGDPpcProvider
+
+    Examples
+    --------
+    >>> provider = GDPpcProvider('low', 'SSP3')  # Requires setting IMPERICS_SHAREDDIR.
+    >>> gdppcs = provider.get_timeseries('ZWE.2.2')  # Get the series for any hierid or ISO
+
+    See Also
+    --------
+    read_bestgdppcprovider : Read files on disk to create a BestGDPpcProvider instance
+    BestGDPpcProvider : Provider of GDP per capita (GDPpc) timeseries, selecting "best" data source
+    """
+    warn(
+        "GDPpcProvider is deprecated, please use read_bestgdppcprovider or BestGDPpcProvider, directly",
+        DeprecationWarning
+    )
+    out = read_bestgdppcprovider(
+        iam=iam,
+        ssp=ssp,
+        growth_path_or_buffer=growth_filepath,
+        baseline_path_or_buffer=baseline_filepath,
+        nightlights_path_or_buffer=nightlights_filepath,
+        use_sharedpath=True,
+        startyear=baseline_year,
+        stopyear=stopyear
+    )
+    return out
+
+
+def _get_best_iso_available(iso, df_this, df_anyiam, df_global):
```

Comment:
It seems cluttering to put this in the module-scope. This function only exists as a step in the `BestGDPpcProvider` algorithm, so it should be bundled as a static method with that object.

Reply:
@jrising You're cluttering your class interface with logic that could have been a simple function. Don't use static methods when a dumb function will do.

Reply:
Imagine that someone in the future wants to read this file and try to understand it. If you just drop this here, they have to figure out where it's used. If in the future there are more implementations of `GDPpcProvider` defined here, they won't know if they can change this without trying to understand the logic of all of the other classes in the file.

One of the biggest reasons for OOP is encapsulation. And all of python is OO, so you're always choosing a level of encapsulation. Here, you've chosen to associate this function with a wider scope than necessary.

Reply:
Python being object-orient doesn't mean that it's good design to make everything a bunch of god classes. It just means that everything is an object (including functions).

Choosing to do this with a a static method or a function doesn't change the level of encapsulation because either one is abstracted away from how users interact with the class.

On one has you have someone like Luciano Ramalho who argues that static method is completely useless, on the other hand you have someone like Leonardo Rochael, who argues that it's useful because you can have child classes modify behavior (I'd agree that this is a useful case). Nobody argues that a python best practice is to make a static method simply because it has some relation to the object. This is not a good practice, ever.

Reply:
Addressed this in `d774eb3e`. This removes the `_get_best_iso_available()` function tests and sticks `_get_best_iso_available()` into an instance method. I went with instance method on this one because it felt a bit more clear that it was run from the instantiated object and never called directly from the class itself.